### PR TITLE
Fix Modern Client + Legacy Server Walls

### PR DESCRIPTION
**Pull request by Axionize**
Closes: https://github.com/GrimAnticheat/Grim/pull/1107 https://github.com/GrimAnticheat/Grim/issues/928 https://github.com/GrimAnticheat/Grim/issues/1459

Also Improves performance of DynamicCollisionWall.fetch() measurably

### CODE REVIEW

File: src/main/java/ac/grim/grimac/utils/collisions/blocks/connecting/DynamicCollisionWall.java [hunk 1]
```diff
@@ -115,60 +117,64 @@ public CollisionBox fetchRegularBox(GrimPlayer player, WrappedBlockState state,
         return new SimpleCollisionBox(f, 0.0F, f2, f1, 1, f3);
     }
 
+    /*
+        This implementation together with the simulation engine ave some limitations.
+        Running into/being knocked into corner walls on a legacy server on a modern client and vice versa
+        Lead to simulation falses. Fixing this rare edge case requires lots more effort than worth and is low priority
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/utils/collisions/blocks/connecting/DynamicCollisionWall.java]:
```suggestion
     * This implementation together with the simulation engine have some limitations.
     * Running into/being knocked into corner walls on a legacy server on a modern client and vice versa
     * Lead to simulation falses. Fixing this rare edge case requires lots more effort than worth and is low priority
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/DynamicCollisionBox.java [hunk 2]
```diff
@@ -23,7 +23,7 @@ public DynamicCollisionBox(GrimPlayer player, ClientVersion version, CollisionFa
 
     // Untested but currently unused
     // *should* work because every single one of these eventually becomes a Complex, Simple, or NoCollision Box
-    @Override
+//    @Override
```
Comment (SamB440) [inline src/main/java/ac/grim/grimac/utils/collisions/datatypes/DynamicCollisionBox.java:26]: Why!?

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java [hunk 3]
```diff
@@ -164,8 +164,7 @@ public Vector[] corners() {
         return vectors;
     }
 
-    @Override
-    public CollisionBox union(SimpleCollisionBox other) {
+    public CollisionBox encompass(SimpleCollisionBox other) {
```
Comment (SamB440) [inline src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:167]: Why change the name?

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java [hunk 4]
```diff
@@ -214,6 +213,11 @@ public SimpleCollisionBox combineToMinimum(double x, double y, double z) {
         return this;
     }
 
+    @Override
+    public CollisionBox union(SimpleCollisionBox other) {
```
Comment (SamB440) [inline src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:217]: Huh? What’s this do now?

**Review by SamB440 — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java (hunk 4 above)
Comment (Axionize) [reply src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:217]: union() actually unions the boxes into an object that includes both the existing and other box. This is different from encompass() which is SCB specific and combines the two SCBs into one new SCB that encompasses both (and potentially more space)

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/DynamicCollisionBox.java (hunk 2 above)
Comment (Axionize) [reply src/main/java/ac/grim/grimac/utils/collisions/datatypes/DynamicCollisionBox.java:26]: I'll fix

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java (hunk 3 above)
Comment (Axionize) [reply src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:167]: It is no longer the same method. See below.

File: src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java (hunk 4 above)
Comment (SamB440) [reply src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:217]: Why change the push handler to use encompass then?

Comment (SamB440) [reply src/main/java/ac/grim/grimac/utils/collisions/datatypes/SimpleCollisionBox.java:217]: Nvm am dum

**Review by SamB440 — APPROVED** [review]

**Review by ManInMyVan — APPROVED** [review]

**Comment by ghost** [comment]
wall collision is also broken for modern clients on modern version (no viaversion) see #1884

**Pull request merged**